ezyhttp-client/src/main/java/com/tvd12/ezyhttp/client/HttpClientProxy.java

```diff
@@ -23,190 +15,191 @@
-	
-	public ResponseEntity request(Request request, int timeout) throws Exception {
-		EzyFuture future = new EzyFutureTask();
-		futures.addFuture(request, future);
-		try {
-			addRequest(request);
-		}
-		catch (Exception e) {
-			futures.removeFuture(request);
-			throw e;
-		}
-		ResponseEntity response = future.get(timeout);
-		return response;
-	}
-	
-	@SuppressWarnings({"rawtypes", "unchecked"})
-	public void fire(Request request, RequestCallback callback) {
-		execute(request, new RequestCallback<ResponseEntity>() {
-			@Override
-			public void onResponse(ResponseEntity response) {
-				try {
-					Object body = response.getBody();
-					callback.onResponse(body);
-				}
-				catch (Exception e) {
-					onException(e);
-				}
-			}
-			@Override
-			public void onException(Exception e) {
-				callback.onResponse(e);
-			}
-		});
-	}
-	
-	public void execute(Request request, RequestCallback<ResponseEntity> callback) {
-		EzyFuture future = new RequestFutureTask(callback);
-		futures.addFuture(request, future);
-		try {
-			addRequest(request);
-		}
-		catch (Exception e) {
-			futures.removeFuture(request);
-			throw e;
-		}
-	}
-	
-	protected void addRequest(Request request) {
-		if(!active)
-			throw new ClientNotActiveException();
-		if(!requestQueue.add(request))
-			throw new RequestQueueFullException(requestQueue.getCapacity());
-	}
-	
-	public static Builder builder() {
-		return new Builder();
-	}
-	
-	public static class Builder implements EzyBuilder<HttpClientProxy> {
-		protected int threadPoolSize;
-		protected int requestQueueCapacity;
-		protected HttpClient.Builder clientBuilder;
-		
-		public Builder() {
-			this.threadPoolSize = 16;
-			this.requestQueueCapacity = 10000;
-			this.clientBuilder = HttpClient.builder();
-			
-		}
-		
-		public Builder readTimeout(int readTimeout) {
-			clientBuilder.readTimeout(readTimeout);
-			return this;
-		}
-		
-		public Builder connectTimeout(int connectTimeout) {
-			clientBuilder.connectTimeout(connectTimeout);
-			return this;
-		}
-		
-		public Builder setStringConverter(Object converter) {
-			clientBuilder.setStringConverter(converter);
-			return this;
-		}
-		
-		public Builder addBodyConverter(Object converter) {
-			clientBuilder.addBodyConverter(converter);
-			return this;
-		}
-		
-		public Builder addBodyConverters(List<?> converters) {
-			clientBuilder.addBodyConverters(converters);
-			return this;
-		}
-		
-		public Builder threadPoolSize(int threadPoolSize) {
-			this.threadPoolSize = threadPoolSize;
-			return this;
-		}
-		
-		protected Builder requestQueueCapacity(int requestQueueCapacity) {
-			this.requestQueueCapacity = requestQueueCapacity;
-			return this;
-		}
-		
-		@Override
-		public HttpClientProxy build() {
-			HttpClientProxy proxy = new HttpClientProxy(
-					threadPoolSize,
-					requestQueueCapacity,
-					clientBuilder.build()
-			);
-			return proxy;
-		}
-	}
-	
+import java.util.List;
+import java.util.Map;
+
+public class HttpClientProxy<T, R>
```

Comment:
a client can serve any type of request response (it means a client will be a [singleton object](https://en.wikipedia.org/wiki/Singleton_pattern) , so I can not specific a generic type for it.
You will get the error:

```
Type mismatch: cannot convert from HttpClientProxy<Object,Object> to HttpClientProxy<HttpClientTest.LoginRequest,HttpClientTest.LoginResponse>
```
If you using

```java
HttpClientProxy<LoginRequest, LoginResponse> client = HttpClientProxy.builder()
				.build();
```

You must always use:

```java
HttpClientProxy<?, ?> client = HttpClientProxy.builder()
				.build();
```

so `<?, ?>` is excessive and I mean generics type is unnecessary

Reply:
@tienampham23 don't use something like that

```java
HttpClientProxy<LoginRequest, LoginResponse> client = new HttpClientProxy<>(HttpClientProxy.builder());
```

because you can see you duplicate word `HttpClientProxy`, it make your source is complicated, you must use 

```java
HttpClientProxy.builder().build();
```

In the future I will change constructor from public to protected to help some avoid create `new HttpClientProxy<>(HttpClientProxy.builder());`

Reply:
@tvd12 I got it, thank you (y)

Reply:
In fact, HttpClientProxy will call any type of request like, `LoginRequest`, `RegisterRequest` so it can not have a generic type